### PR TITLE
Ability to create new client for the API

### LINKED ISSUE #1597: [v2] Client ID/secret
Feels weird to ask this in a bug report, but irc-channel seems pretty dead and also got no reply to the mail I sent to hello@wallabag.org. And since there was this statement at the end of the oAuth2 blog-post
> If you have some problems with the API, feel free to create a new issue on GitHub.

So I want to support a "share article to wallabag" feature in my RSS-Reader. How do I get a client_id and client_secret for this?
Your blog-post says:
> As we are still in alpha, we create a client for you (but in few weeks, each application will have its own client id and client secret).

But the tokens in the post didn't work anymore
>client_id=1_3bcbxd9e24g0gk4swg0kwgcwg4o8k8g4g888kwc44gcc0gwwk4
>client_secret=4ok2x70rlfokc8g0wws8c8kwcokw80k44sg48goc0ok4w0so0k

results in 

> "The client credentials are invalid"

So how do I test my code with v2.wallabag.org?
thx

### PR DESCRIPTION
**Pull request by nicosomb**
Fix #1597 
Just a first draft. 

Todo : 

- [x] a form to create a new client (today, each time we click on "create a new client", we create a new client ...)
- [x] minor enhancements
- [x] tests
- [x] translations

### CODE REVIEW

**Comment by tcitworld** [comment]
Oh, that reminds me we wanted to use prismjs for code coloration inside content. TODO.

File: src/Wallabag/CoreBundle/Controller/DeveloperController.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    public function createClientAction(Request $request)
+    {
+        $em = $this->getDoctrine()->getManager();
+        $client = new Client();
+        $clientForm = $this->createForm(ClientType::class, $client);
+        $clientForm->handleRequest($request);
+
+        if ($clientForm->isValid()) {
+            $client->setAllowedGrantTypes(array('token', 'authorization_code'));
```
Comment (tcitworld) [inline src/Wallabag/CoreBundle/Controller/DeveloperController.php]: Shouldn't there be `password` too here ?

Comment (nicosomb) [inline src/Wallabag/CoreBundle/Controller/DeveloperController.php]: I don't know. Before this PR, I followed this gist https://gist.github.com/tjamps/11d617a4b318d65ca583 and there is only password. I need to read some documentation about that. @j0k3r any opinion?

**Comment by tcitworld** [comment]
Could a client also be added through a command, something like 
```
wallabag:oauth-server:client:create --redirect-uri="myappURLcallback" --grant-type="authorization_code" --grant-type="password" --grant-type="refresh-token" --grant-type="token" --grant-type="client_credentials"
```

File: src/Wallabag/CoreBundle/Tests/Controller/DeveloperControllerTest.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: nicosomb
+ * Date: 16/02/16
+ * Time: 19:39.
+ */
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Tests/Controller/DeveloperControllerTest.php]: We don't need that :)

File: src/Wallabag/CoreBundle/Controller/DeveloperController.php (hunk 1 above)
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Controller/DeveloperController.php]: I don't think we have to allow password here.
(jump to password section) https://aaronparecki.com/2012/07/29/2/oauth2-simplified

**Comment by j0k3r** [comment]
Could you add some little tests to the developer & howto page?

File: src/Wallabag/CoreBundle/Controller/DeveloperController.php (hunk 1 above)
Comment (tcitworld) [inline src/Wallabag/CoreBundle/Controller/DeveloperController.php]: So, we don't allow auth through password ?

File: src/Wallabag/CoreBundle/Tests/Controller/DeveloperControllerTest.php (hunk 2 above)
Comment (nicosomb) [inline src/Wallabag/CoreBundle/Tests/Controller/DeveloperControllerTest.php]: :open_mouth:

**Comment by nicosomb** [comment]
> Could you add some little tests to the developer & howto page?

Just to test if pages are well displayed?

**Comment by j0k3r** [comment]
> Just to test if pages are well displayed?

Yep, return 200 and contains something we know

**Comment by j0k3r** [comment]
I think there is sth to do about the link at the bottom:

capture d ecran 2016-02-19 a 10 47 29

And how can we find the link to developper when using the baggy theme?

**Comment by j0k3r** [comment]
I've added a listing of clients which is collapsed by default:

capture d ecran 2016-03-05 a 21 44 56

When opened:

capture d ecran 2016-03-05 a 21 45 02

File: src/Wallabag/CoreBundle/Resources/translations/messages.fr.yml [hunk 3]
```diff
@@ -240,3 +240,36 @@ If you need some help, we are here for you.: "Parce que vous avez peut-être bes
...
+Client secret: "Clé secrète"
+Redirect URIs: "URLs de redirection"
+Grant type allowed: "Type de privilège accordé"
+You have the ability to remove this client. This action is IRREVERSIBLE !: "Vous avez la possibilité de supprimer un client. Cette action est IRREVERSIBLE !"
+If you remove it, every app configured with that client won't be able to auth on your wallabag.: "Si vous supprimez un client, toutes les applications qui l'utilisaient ne fonctionneront plus avec votre compte wallabag."
+Remove this client: "Supprimer ce client"
+New client: "Nouveau client"
+You are about to create a new client. Please fill the field below for the redirect URI of your application.: "Vous allez créer un nouveau client. Merci de remplir l'url de redirection vers votre application."
+Back: "Retour"
+Client parameters: "Les parameters de votre client"
```
Comment (nicosomb) [inline src/Wallabag/CoreBundle/Resources/translations/messages.fr.yml]: parameters > paramètres

File: src/Wallabag/CoreBundle/Resources/translations/messages.fr.yml [hunk 4]
```diff
@@ -240,3 +240,36 @@ If you need some help, we are here for you.: "Parce que vous avez peut-être bes
...
+New client created.: "Nouveau client créé."
+Here are your client parameters.: "Voilà les paramètres de votre client"
+Read the howto "Create my first application": "Lire \"comment créer ma première application\""
+Client deleted: "Client supprimé"
+No client yet.: "Aucun client pour le moment"
+"The following commands make use of the <a href=\"https://github.com/jkbrzt/httpie\">HTTPie library</a>. Make sure it is installed on your system before using it.": "Les commandes suivantes utilisent la <a href=\"https://github.com/jkbrzt/httpie\">librarie HTTPie</a>. Assurez-vous qu'elle soit installée avant de l'utiliser."
+You need a token to communicate between your 3rd application and wallabag API.: "Vous avez besoin d'un token pour échanger entre votre appication et l'API de wallabag."
```
Comment (nicosomb) [inline src/Wallabag/CoreBundle/Resources/translations/messages.fr.yml]: appication > application

File: src/Wallabag/CoreBundle/Resources/translations/messages.fr.yml [hunk 5]
```diff
@@ -240,3 +240,36 @@ If you need some help, we are here for you.: "Parce que vous avez peut-être bes
...
+"To create this token, you need <a href=\"%link%\">to create a new client</a>.": "Pour créer un token, vous devez <a href=\"%link%\">créer un nouveau client</a>."
+Now, create your token (replace client_id, client_secret, username and password with the good values):: "Maintenant créer votre token (remplacer client_id, client_secret, username et password avec les bonnes valeurs):"
```
Comment (nicosomb) [inline src/Wallabag/CoreBundle/Resources/translations/messages.fr.yml]: créer > créez

File: src/Wallabag/CoreBundle/Resources/translations/messages.fr.yml [hunk 6]
```diff
@@ -240,3 +240,36 @@ If you need some help, we are here for you.: "Parce que vous avez peut-être bes
...
+The API will return a response like this:: "L'API vous retournera une réponse comme ça:"
+The access_token is useful to do a call to the API endpoint. For example:: "L'access_token doit être utilisé pour faire un appel à l'API. Par exemple:"
```
Comment (nicosomb) [inline src/Wallabag/CoreBundle/Resources/translations/messages.fr.yml]: `exemple:` > `exemple :`

File: src/Wallabag/CoreBundle/Resources/translations/messages.fr.yml [hunk 7]
```diff
@@ -240,3 +240,36 @@ If you need some help, we are here for you.: "Parce que vous avez peut-être bes
...
+This call will return all the entries for your user.: "Cet appel va retourner tous les articles de l'utilisateur."
+"If you want to see all the API endpoints, you can have a look <a href=\"%link%\">to our API documentation</a>.": "Si vous voulez toutes les méthodes de l'API, jettez un oeil <a href=\"%link%\">à la documentation de l'API</a>."
```
Comment (nicosomb) [inline src/Wallabag/CoreBundle/Resources/translations/messages.fr.yml]: jettez > jetez

File: src/Wallabag/CoreBundle/Resources/views/themes/material/layout.html.twig [hunk 8]
```diff
@@ -116,6 +116,8 @@
             <div class="container">
                 <p>{% trans %}powered by{% endtrans %} <a target="_blank" href="https://wallabag.org" class="grey-text text-lighten-4">wallabag</a></p>
                 <a class="grey-text text-lighten-4 right" href="{{ path('about') }}">{% trans %}About{% endtrans %}</a>
+                <a class="grey-text text-lighten-4 right" href="{{ path('developer') }}">{% trans %}Developer{% endtrans %}
```
Comment (nicosomb) [inline src/Wallabag/CoreBundle/Resources/views/themes/material/layout.html.twig]: We can remove this line if we add a new entry in the menu.

Comment (j0k3r) [inline src/Wallabag/CoreBundle/Resources/views/themes/material/layout.html.twig]: The menu entry is only for baggy, I'll add it to material too.

File: src/Wallabag/CoreBundle/Resources/views/themes/material/layout.html.twig [hunk 9]
```diff
@@ -50,6 +50,7 @@
             {% endif %}
             <li class="bold {% if currentRoute == 'import' %}active{% endif %}"><a class="waves-effect" href="{{ path('import') }}">{% trans %}import{% endtrans %}</a></li>
             <li class="bold {% if currentRoute == 'howto' %}active{% endif %}"><a class="waves-effect" href="{{ path('howto') }}">{% trans %}howto{% endtrans %}</a></li>
+            <li class="bold {% if currentRoute == 'developer' %}active{% endif %}"><a class="waves-effect" href="{{ path('developer') }}">{% trans %}Developer{% endtrans %}</a></li>
```
Comment (nicosomb) [inline src/Wallabag/CoreBundle/Resources/views/themes/material/layout.html.twig]: Developer > developer, or we need to change all the menu to add capitalized letter maybe.

Comment (j0k3r) [inline src/Wallabag/CoreBundle/Resources/views/themes/material/layout.html.twig]: This'll be fixed when https://github.com/wallabag/wallabag/issues/1697 will be implemented.

**Comment by nicosomb** [comment]
:+1: good job!

**Comment by tcitworld** [comment]
I would have preferred if we could have the possibility to set up a name for each client. :arrow_right: #1765

**Pull request merged**